---
title: Species tree inference with ASTRAL
---

In this page, we will see how to use the trees calculated at the previous step to infer a _species tree_. In this context, _species tree_ refers to a phylogeny representing the history of splits across individuals in the dataset, as opposed to a _gene tree_, which represents the evolutionary relationships of the DNA sequences carried by these individuals at a given locus of the genome. In that sense, the sliding windows trees that we inferred earlier are _gene trees_, with loci defined by the windows. The distinction between _species tree_ and _gene tree_ is important, because the relationship at any given locus may be discordant from the actual evolutionary history of the individuals, because of variations in the order of coalescence events that can be caused for exemple by introgression or incomplete lineage sorting. Despite the name, a _species tree_ does not necessarily describe relationships between species, and the tips can represent any taxonomic level. Assuming an absence of introgression, the _species tree_ is a function of the levels of variation across _gene trees_, and several methods have been developped to estimate the former from the latter (in the presence of introgression things get trickier, but we will talk about it in another page). Here we will use a software called ASTRAL. More details about species trees and related softwares can be found in the litterature.

